numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3423,6 +3428,43 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
         /* Get the result from the iterator object array */
         ret = (PyObject*)NpyIter_GetOperandArray(iter)[0];
 
+        PyArray_Descr **dts = NpyIter_GetDescrArray(iter);
+        PyArray_Descr *dtx = dts[2];
+        PyArray_Descr *dty = dts[3];
+        npy_intp itemsize = dts[0]->elsize;
+
+        npy_intp *strides = NpyIter_GetInnerStrideArray(iter);
+        npy_intp cstride = strides[1];
+        npy_intp xstride = strides[2];
+        npy_intp ystride = strides[3];
+
+        int axswap = PyDataType_ISBYTESWAPPED(dtx);
+        int ayswap = PyDataType_ISBYTESWAPPED(dty);
+        int native = (axswap == ayswap) && (axswap == 0) && !needs_api;
+
+        NPY_ARRAYMETHOD_FLAGS x_transfer_flags;
+        NPY_ARRAYMETHOD_FLAGS y_transfer_flags;
+
+        int x_is_aligned = IsAligned(ax);
+        int y_is_aligned = IsAligned(ay);
+
+        npy_intp xstrides[2] = {xstride, itemsize};
+        npy_intp ystrides[2] = {xstride, itemsize};
+
+        npy_intp one = 1;
+
+        if (PyArray_GetDTypeTransferFunction(
+                x_is_aligned, xstrides[0], xstrides[1], dtx, common_dt, 0,
+                &x_cast_info, &x_transfer_flags) != NPY_SUCCEED) {
+            goto fail;
+        }
+
+        if (PyArray_GetDTypeTransferFunction(
+                y_is_aligned, ystrides[0], ystrides[1], dty, common_dt, 0,
+                &y_cast_info, &y_transfer_flags) != NPY_SUCCEED) {
+            goto fail;
+        }
```

Comment:
Theoretically a point if one of them is not aligned, but the iterator probably even ensures that.

We really only need a single transfer function here, I think.  The old code is unnecessarily complex: the iterator ensures that swapping cannot possibly be necessary.  (may not be always ideal, but that is what it does.)